# Career Foundry - Achievement 6 - Task 1

# Table of Contents
## Import Libraries and Data
## Step 6 - Data Cleaning and Consistency Checks
### Explore Data
### Check for Mixed-Type Data
### Check for Missing Values
### Check for Duplicate Values
### Correct Issues in Birth Year Data
## Step 7 - Descriptive Statistical Analysis

# Import Libraries and Data

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Import data
path = r"C:\Users\carol\Documents\Career Foundry\6 Advanced Analytics & Dashboard Design\A6 Project"
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'citibike.csv'), index_col = False)

# Step 6 - Data Cleaning and Consistency Checks

## Explore Data

In [3]:
df.head()

trip_id  bike_id weekday  start_hour           start_time  start_station_id  \
0  LnQzQk    16013     Mon          18  2013-09-09 18:18:55               523   
1  IL9boN    15230     Thu          18  2013-09-12 18:38:53               257   
2  46clGB    17942     Wed          19  2013-09-18 19:44:04               479   
3  v7vdFt    19683     Sat          11  2013-09-28 11:54:37               527   
4  VGBsb5    18024     Sat          18  2013-09-07 18:08:22               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

              end_time  end_station_id     end_station_name  \
0  2013-09-09 18:35:28             334      W 20 St & 7 Ave   
1  2013-09-12 18:48:34             236  St Marks Pl & 2 Ave   
2  2013-09-18 19:50:05             513     W 56 St & 10 Ave   
3  2013-09-28 12:03:58             441      E 52 St & 2 Ave   
4  2013-09-07 18:46:38             476      E 31 St & 3 Ave   

   end_station_latitude  end_station_longitude  trip_duration      subscriber  \
0             40.742388             -73.997262            993      Subscriber   
1             40.728419             -73.987140            581      Subscriber   
2             40.768254             -73.988639            361      Subscriber   
3             40.756014             -73.967416            561      Subscriber   
4             40.743943             -73.979661           2296  Non-Subscriber   

   birth_year  gender  
0      1968.0       2  
1      1983.0       1  
2      1989.0       1  
3      1988.0       2  
4         NaN       0

In [4]:
df.describe()

bike_id    start_hour  start_station_id  start_station_latitude  \
count  50000.000000  50000.000000      50000.000000            50000.000000   
mean   17615.269360     14.145240        443.321500               40.734170   
std     1675.407446      4.860541        356.559925                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       start_station_longitude  end_station_id  end_station_latitude  \
count             50000.000000    50000.000000          50000.000000   
mean                -73.991109      442.539700             40.733859   
std                   0.012555      355.756022              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       end_station_longitude  trip_duration    birth_year        gender  
count           50000.000000   50000.000000  43021.000000  50000.000000  
mean              -73.991351     838.982900   1975.627786      1.073540  
std                 0.012569     573.663997     11.089001      0.589389  
min               -74.017134      60.000000   1899.000000      0.000000  
25%               -74.001547     417.000000   1968.000000      1.000000  
50%               -73.991218     672.000000   1978.000000      1.000000  
75%               -73.982050    1112.000000   1984.000000      1.000000  
max               -73.950048    2697.000000   1997.000000      2.000000

Most of these numbers make sense. Based on the values in trip duration, I think the trip duration is most likely recorded in seconds. There is almost certainly an error in the birth year values, because it lists the minimum as 1899 and I highly doubt someone over 100 has used Citibike. I'll also need to figure out the meaning of the values used in gender, i.e. whether 1 = male or female, and whether 0 is a value used for people that marked an option like "prefer not to answer" or if it means non-binary.

## Check for Mixed-Type Data

In [5]:
#Check for mixed types
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

There are no columns with mixed type data in the results dataframe

## Check for Missing Values

In [6]:
#Check each column for missing values and sum them
df.isnull().sum()

trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
dtype: int64

Birth year has 6,979 missing values.

## Check for Duplicate Values

In [7]:
#Check for full duplicates in the dataframe
df_dups = df[df.duplicated()]

In [8]:
#Print duplicates dataframe to see if there are any duplicates
df_dups

Empty DataFrame
Columns: [trip_id, bike_id, weekday, start_hour, start_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_time, end_station_id, end_station_name, end_station_latitude, end_station_longitude, trip_duration, subscriber, birth_year, gender]
Index: []

There are no full duplicate values in the results dataframe

## Correct Issues in Birth Year Data

Birth year is the only column that seems to have any issues (6979 missing values and a minimum year of 1899 which is very unlikely). 

In [18]:
df['birth_year'].value_counts(dropna = False)

NaN       6979
1985.0    1837
1981.0    1830
1982.0    1810
1984.0    1798
1983.0    1785
1986.0    1714
1980.0    1645
1979.0    1537
1987.0    1530
1988.0    1503
1978.0    1384
1989.0    1302
1977.0    1269
1976.0    1261
1974.0    1191
1975.0    1155
1970.0    1131
1969.0    1054
1972.0    1046
1973.0    1018
1971.0     969
1990.0     932
1968.0     888
1967.0     888
1966.0     833
1963.0     761
1965.0     738
1964.0     691
1962.0     677
1960.0     622
1961.0     554
1959.0     534
1991.0     473
1957.0     462
1958.0     431
1956.0     398
1992.0     395
1955.0     349
1953.0     321
1993.0     314
1954.0     311
1952.0     239
1951.0     209
1950.0     179
1994.0     149
1948.0     147
1949.0     130
1947.0      96
1995.0      82
1996.0      67
1946.0      60
1945.0      57
1944.0      41
1942.0      37
1943.0      35
1997.0      26
1940.0      20
1941.0      18
1935.0      14
1932.0      10
1900.0       9
1899.0       8
1939.0       8
1937.0       7
1938.0       7
1933.0    

In [19]:
df[df['birth_year'] < 1933].count()

trip_id                    41
bike_id                    41
weekday                    41
start_hour                 41
start_time                 41
start_station_id           41
start_station_name         41
start_station_latitude     41
start_station_longitude    41
end_time                   41
end_station_id             41
end_station_name           41
end_station_latitude       41
end_station_longitude      41
trip_duration              41
subscriber                 41
birth_year                 41
gender                     41
dtype: int64

There are 41 records with an age greater than 80 (the data goes up to 2013). This accounts for 0.082% of the total data. Because it accounts for such a small portion of the data, I'm going to leave those years alone. 

In [20]:
df['gender'].value_counts(dropna = False)

1    32361
2    10658
0     6981
Name: gender, dtype: int64

In [ ]:
df[df['gender'] == 0]

In [22]:
df['subscriber'].value_counts(dropna = False)

Subscriber        43021
Non-Subscriber     6979
Name: subscriber, dtype: int64

I dug deeper into the data and found that the data with missing birth years corresponds to non-subscribers (these records also have gender = 0). This makes sense since non-subscibers would probably not be required to enter personal information like that. Since these records already have a non-subscriber flag I'm going to leave them alone for now.

# Step 7 - Descriptive Statistical Analysis

In [23]:
#Export descriptive statistical analysis for results data
dsa = df.describe()
dsa.to_clipboard()